### ISSUE TITLE
Clickhouse shard data disappearance

### ISSUE THREAD

**Issue by igor-sh8** [ClickHouse/ClickHouse#3304]
Hello. 

Faced such a problem in clickhouse. 
About a month ago (beginning of September) we created a cluster (3 shards, 2 replicas in each shard). On all clickhouse servers we created a ReplicatedSummingMergeTree table like this: 
`CREATE TABLE Table1 ( Date Date,  Col1 Int32, Col2 In32, ... ,Col10 Int32) ENGINE = ReplicatedSummingMergeTree('/clickhouse/tables/{shard}/Table1', '{replica}', Date, (Col1, Col2, ... ,Col5), 8192, (Col6, ... ,Col10))`
and distributed table like this:
`CREATE TABLE Table1_Distr ( Date Date,  Col1 Int32, Col2 In32, ... Col10 Int32) ENGINE = Distributed(cluster_name, default, Table1, cityHash64(Col1, Col2, ... ,Col5))`

In September all worked fine, but on October 4 we noticed that the statistics for the first days of October partially disappeared on one of the shards  (1st and 3rd days partially disappeared and 2nd day disappear completely). With statistics for the 4th of October all fine.
Interesting thing that the statistics for the 1-3 days of October has been saved (we saw that statistics) but then just disappeared.
Checked all clickhouse logs - all clear there (information log level).
What could it be ? 
Could this be some kind of unsuccessful automatic resharding?

**Comment by den-crane** [comment]
Hi, one of the possible reason that you did not include date in your sort-key.

```
CREATE TABLE Table1 ( date Date, col1 UInt64, col2 Float64) ENGINE = SummingMergeTree(date, col1, 8192)
insert into Table1 values('2018-10-01', 1, 6);
insert into Table1 values('2018-10-02', 1, 6);
insert into Table1 values('2018-10-10', 1, 6);
optimize table Table1 final;
select * from Table1
┌───────date─┬─col1─┬─col2─┐
│ 2018-10-01 │    1 │   18 │
└────────────┴──────┴──────┘
```

Now sort-key = (col1, date)

```
CREATE TABLE Table1 ( date Date, col1 UInt64, col2 Float64) ENGINE = SummingMergeTree(date, (col1, date), 8192)
insert into Table1 values('2018-10-01', 1, 6);
insert into Table1 values('2018-10-02', 1, 6);
insert into Table1 values('2018-10-10', 1, 6);
optimize table Table1 final;
select * from Table1
┌───────date─┬─col1─┬─col2─┐
│ 2018-10-01 │    1 │    6 │
│ 2018-10-02 │    1 │    6 │
│ 2018-10-10 │    1 │    6 │
└────────────┴──────┴──────┘
```

**Comment by den-crane** [comment]
There is no any automatic resharding in CH.

Old syntax (partition by month): SummingMergeTree(date, (col1,date), 8192)
New syntax (partition by month): SummingMergeTree partition by toYYYYMM(date), order by (col1,date)

New syntax (partition by day): SummingMergeTree partition by (date), order by (col1,date)
and with partition by day you can use SummingMergeTree partition by (date), order by (col1)

**Comment by igor-sh8** [comment]
Oh, sorry. My fault. I made a mistake in the create table statement. 
Date column already included in the sort-key. 
`CREATE TABLE Table1 ( Date Date, Col1 Int32, Col2 In32, ... ,Col10 Int32) ENGINE = ReplicatedSummingMergeTree('/clickhouse/tables/{shard}/Table1', '{replica}', Date, (Date, Col1, Col2, ... ,Col5), 8192, (Col6, ... ,Col10))`

**Comment by igor-sh8** [comment]
Hi. 
Maybe our problem looks like this bug fix description in 18.14.9 version:
`Если после слияния кусков данных, у результирующего куска чексумма отличается от результата того же слияния на другой реплике, то результат слияния удаляется, и вместо этого кусок скачивается с другой реплики (это правильное поведение). Но после скачивания куска, он не мог добавиться в рабочий набор из-за ошибки, что кусок уже существует (так как кусок после слияния удалялся не сразу, а с задержкой). Это приводило к циклическим попыткам скачивания одних и тех же данных.` 
? 
Maybe we had an error at the stage of cyclic download attempts?
Could this happen?

**Comment by alexey-milovidov** [comment]
No, it could not happen.
Replacing of data parts to the merged or fetched part is completely atomic.
If merge or fetch of data part has failed, the set of data parts will remain in previous state.
